### PR TITLE
✨ Check duplication before inserting records

### PR DESCRIPTION
**Pull request by sunnyosun**

### CODE REVIEW

File: lamindb/_delete.py [hunk 1]
```diff
@@ -16,41 +36,58 @@ def delete(record: sqm.SQLModel):
 
     >>> # Delete metadata records
     >>> experiment = ln.select(Experiment, id=experiment_id)
-    >>> db.delete(experiment)
+    >>> ln.delete(experiment)
     >>> # Delete data objects
     >>> dobject = ln.select(DObject, id=dobject_id)
-    >>> db.delete(dobject)
+    >>> ln.delete(dobject)
+    >>> # Delete without selecting a record
```
Comment (falexwolf) [inline lamindb/_delete.py:43]: How about `# Delete by field`

File: lamindb/_delete.py [hunk 2]
```diff
@@ -16,41 +36,58 @@ def delete(record: sqm.SQLModel):
 
     >>> # Delete metadata records
     >>> experiment = ln.select(Experiment, id=experiment_id)
```
Comment (falexwolf) [inline lamindb/_delete.py:38]: `Delete by record`

File: lamindb/_delete.py (hunk 1 above)
Comment (sunnyosun) [reply lamindb/_delete.py:43]: ok!

File: lamindb/dev/db/_add.py [hunk 3]
```diff
@@ -38,22 +48,34 @@ def add(  # type: ignore  # no support of different naming of args across overlo
...
     >>> # insert a new record
-    >>> db.add(wetlab.experiment(name="My test", biometa_id=test_id))
+    >>> ln.add(wetlab.Experiment(name="My test", biometa_id=test_id))
     >>> # update an existing record
-    >>> experiment = ln.select(wetlab.experiment, id=experiment_id).one()
+    >>> experiment = ln.select(wetlab.Experiment, id=experiment_id).one()
     >>> experiment.name = "New name"
-    >>> db.add(experiment)
+    >>> ln.add(experiment)
+    >>> # check existence before inserting a new record
```
Comment (falexwolf) [inline lamindb/dev/db/_add.py:56]: How about `# add a record for fields if not yet exists`?

Comment (sunnyosun) [reply lamindb/dev/db/_add.py:56]: ok

File: lamindb/dev/db/_add.py [hunk 4]
```diff
@@ -38,22 +48,34 @@ def add(  # type: ignore  # no support of different naming of args across overlo
     Example:
 
     >>> # insert a new record
```
Comment (falexwolf) [inline lamindb/dev/db/_add.py:50]: How about, `# add a record (by passing a record)`

Comment (sunnyosun) [reply lamindb/dev/db/_add.py:50]: ok!

Comment (falexwolf) [reply lamindb/dev/db/_add.py:50]: I think it'd be more readable to write it in two lines, making an experiment record, and passing it.

File: lamindb/dev/db/_add.py [hunk 5]
```diff
@@ -38,22 +48,34 @@ def add(  # type: ignore  # no support of different naming of args across overlo
...
-    >>> db.add(wetlab.experiment(name="My test", biometa_id=test_id))
+    >>> ln.add(wetlab.Experiment(name="My test", biometa_id=test_id))
     >>> # update an existing record
-    >>> experiment = ln.select(wetlab.experiment, id=experiment_id).one()
+    >>> experiment = ln.select(wetlab.Experiment, id=experiment_id).one()
```
Comment (falexwolf) [inline lamindb/dev/db/_add.py:53]: Then this line can be skipped.

File: lamindb/dev/db/_add.py (hunk 3 above)
Comment (sunnyosun) [reply lamindb/dev/db/_add.py:56]: by fields?

File: lamindb/dev/db/_add.py (hunk 5 above)
Comment (sunnyosun) [reply lamindb/dev/db/_add.py:53]: why?

File: lamindb/dev/db/_add.py [hunk 6]
```diff
@@ -38,22 +48,34 @@ def add(  # type: ignore  # no support of different naming of args across overlo
...
+        use_fsspec: Whether to use fsspec.
     """
     if isinstance(record, list):
         records = record
-    else:
+    elif isinstance(record, sqm.SQLModel):
         records = [record]
+    else:
+        model = record
+        results = select(model, **fields).all()
```
Comment (falexwolf) [inline lamindb/dev/db/_add.py:69]: Could we do `.one()`, here?

Comment (falexwolf) [reply lamindb/dev/db/_add.py:69]: For this reason: If think you can only "add by field" if you have one unique record in mind.

Comment (sunnyosun) [reply lamindb/dev/db/_add.py:69]: ok!

Comment (falexwolf) [reply lamindb/dev/db/_add.py:69]: This is why our pattern was: 
```
result = select().one_or_none()
if result is None:
    # add
```

Comment (sunnyosun) [reply lamindb/dev/db/_add.py:69]: Yes, but I thought it doesn’t hurt to return all results? [](edited-with-Axolo)

Comment (falexwolf) [reply lamindb/dev/db/_add.py:69]: If the latter is relevant, maybe that's great!

But the downside is that one has an unexpected return type.

Comment (sunnyosun) [reply lamindb/dev/db/_add.py:69]: Yes, that’s what I did [](edited-with-Axolo)

Comment (falexwolf) [reply lamindb/dev/db/_add.py:69]: So, one can't use it as an assert statement, but will always need to check that it's in fact just one record that's returned.

Comment (falexwolf) [reply lamindb/dev/db/_add.py:69]: Hence, I'd rather throw an error than returning a list if there a more return results in the query.

Comment (sunnyosun) [reply lamindb/dev/db/_add.py:69]: but I think the point is to check whether any records with the field exists [](edited-with-Axolo)

Comment (sunnyosun) [reply lamindb/dev/db/_add.py:69]: without caring much about the uniqueness [](edited-with-Axolo)

Comment (falexwolf) [reply lamindb/dev/db/_add.py:69]: With these fundamental additions to the API, I think the more minimalistic and expectable the behavior, the better.

File: lamindb/dev/db/_add.py (hunk 4 above)
Comment (sunnyosun) [reply lamindb/dev/db/_add.py]: ok! [](edited-with-Axolo)

File: lamindb/dev/db/_add.py (hunk 6 above)
Comment (sunnyosun) [reply lamindb/dev/db/_add.py:69]: I'm not sure; merging now, we can discuss!

**Pull request merged**